ql/src/java/org/apache/hadoop/hive/ql/optimizer/SharedWorkOptimizer.java

```diff
@@ -1474,8 +1489,29 @@ private static SharedResult extractSharedOptimizationInfoForRoot(ParseContext pc
         discardableInputOps.addAll(gatherDPPBranchOps(pctx, optimizerCache, discardableOps));
         discardableInputOps.addAll(gatherDPPBranchOps(pctx, optimizerCache, retainableOps,
             discardableInputOps));
-        return new SharedResult(retainableOps, discardableOps, discardableInputOps,
-            dataSize, maxDataSize);
+        bailOut = true;
+      }
+
+      if (bailOut) {
+        // Accumulate InMemoryDataSize of unmerged MapJoin operators.
+        Set<Operator<?>> opsWork1 = findWorkOperators(optimizerCache, currentOp1);
+        for (Operator<?> op : opsWork1) {
+          if (op instanceof MapJoinOperator) {
+            MapJoinOperator mop = (MapJoinOperator) op;
+            dataSize = StatsUtils.safeAdd(dataSize, mop.getConf().getInMemoryDataSize());
+            maxDataSize = mop.getConf().getMemoryMonitorInfo().getAdjustedNoConditionalTaskSize();
+          }
+        }
+        Set<Operator<?>> opsWork2 = findWorkOperators(optimizerCache, currentOp2);
+        for (Operator<?> op : opsWork2) {
+          if (op instanceof MapJoinOperator) {
+            MapJoinOperator mop = (MapJoinOperator) op;
+            dataSize = StatsUtils.safeAdd(dataSize, mop.getConf().getInMemoryDataSize());
+            maxDataSize = mop.getConf().getMemoryMonitorInfo().getAdjustedNoConditionalTaskSize();
+          }
```

Comment:
This logic already exists in lines 1430-1446. Could you please extract it to a method and remove the variable `bailOut` and call that method.